components/TextFields/src/ContainedInputView/MDCBaseTextField.m

```diff
@@ -369,6 +382,37 @@ - (UIFont *)uiTextFieldDefaultFont {
   return font;
 }
 
+#pragma mark Placeholder
+
+- (BOOL)shouldPlaceholderBeVisible {
+  return [self shouldPlaceholderBeVisibleWithPlaceholder:self.placeholder
+                                              labelState:self.labelState
+                                                    text:self.text
+                                               isEditing:self.isEditing];
+}
+
+- (BOOL)shouldPlaceholderBeVisibleWithPlaceholder:(NSString *)placeholder
+                                       labelState:(MDCContainedInputViewLabelState)labelState
+                                             text:(NSString *)text
+                                        isEditing:(BOOL)isEditing {
+  BOOL hasPlaceholder = placeholder.length > 0;
+  BOOL hasText = text.length > 0;
+
+  if (hasPlaceholder) {
+    if (hasText) {
+      return NO;
+    } else {
+      if (labelState == MDCContainedInputViewLabelStateNormal) {
```

Comment:
Yes, and let us add some snapshots for all the branching logic we have put here.
Looks like there are 4 cases to cover here.

Reply:
How about a unit test instead? This looks to me like a good candidate for a unit test.

Reply:
Added a unit test.

Reply:
Sharing testing best practices: If we have to expose the a private method in our tests it is a code smell. It becomes a brittle test if we ever refactor the code. This is why I suggested a snapshot test.

Reply:
Yes, this unit test will probably have to change if we change the conditions under which the placeholder should be visible. I added a snapshot test for a textfield with a placeholder but I didn't add one for every possible code path in the method because I think placeholder visibility behavior can generally be captured by other snapshot tests not necessarily devoted to it.